device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsCbsSenderLinkHandler.java

```diff
@@ -0,0 +1,120 @@
+    private static final String NAME_KEY = "name";
+
+    private static final String OPERATION_VALUE = "put-token";
+    private static final String TYPE_VALUE = "servicebus.windows.net:sastoken";
+
+    private static final String DEVICES_PATH = "/devices/";
+
+    AmqpsCbsSenderLinkHandler(Sender sender, AmqpsLinkStateCallback amqpsLinkStateCallback)
+    {
+        super(sender, amqpsLinkStateCallback, UUID.randomUUID().toString());
+
+        this.senderLinkTag = SENDER_LINK_TAG_PREFIX;
+        this.senderLinkAddress = SENDER_LINK_ENDPOINT_PATH;
+    }
+
+    static String getCbsTag()
+    {
+        return SENDER_LINK_TAG_PREFIX;
+    }
+
+    @Override
+    public String getLinkInstanceType()
+    {
+        return "cbs";
+    }
+
+    UUID sendAuthenticationMessage(DeviceClientConfig deviceClientConfig) throws TransportException
+    {
+        UUID correlationId = UUID.randomUUID();
+        MessageImpl outgoingMessage = createCBSAuthenticationMessage(deviceClientConfig, correlationId);
+
+        AmqpsSendResult sendResult = this.sendMessageAndGetDeliveryTag(outgoingMessage);
+
+        //This message will be ignored when this send is acknowledged, so just provide an empty message for the map
+        inProgressMessages.put(sendResult.getDeliveryTag(), new Message());
+
+        return correlationId;
+    }
+
+    private MessageImpl createCBSAuthenticationMessage(DeviceClientConfig deviceClientConfig, UUID correlationId) throws TransportException
+    {
+        MessageImpl outgoingMessage = (MessageImpl) Proton.message();
+
+        Properties properties = new Properties();
+
+        // Note that setting "messageId = correlationId" is intentional.
+        // IotHub only responds correctly if this correlation id is set this way
+        properties.setMessageId(correlationId);
+
+        properties.setTo(CBS_TO);
+        properties.setReplyTo(CBS_REPLY);
+        outgoingMessage.setProperties(properties);
+
+        Map<String, Object> userProperties = new HashMap<>(3);
+        userProperties.put(OPERATION_KEY, OPERATION_VALUE);
+        userProperties.put(TYPE_KEY, TYPE_VALUE);
+
+        String host = deviceClientConfig.getGatewayHostname();
+        if (host == null || host.isEmpty())
+        {
+            host = deviceClientConfig.getIotHubHostname();
+        }
+
+        userProperties.put(NAME_KEY, host + DEVICES_PATH + deviceClientConfig.getDeviceId());
+        ApplicationProperties applicationProperties = new ApplicationProperties(userProperties);
+        outgoingMessage.setApplicationProperties(applicationProperties);
+
+        Section section;
+        try
+        {
+            section = new AmqpValue(deviceClientConfig.getSasTokenAuthentication().getRenewedSasToken(true, true));
+            outgoingMessage.setBody(section);
+        }
+        catch (IOException e)
+        {
+            log.error("getRenewedSasToken has thrown exception while building new cbs authentication message", e);
```

Comment:
getRenewedSasToken?

Reply:
I'll change this to "Failed to renew sas token while building..."